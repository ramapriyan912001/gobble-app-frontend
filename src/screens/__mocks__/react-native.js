// import * as ReactNative from 'react-native';

// jest.doMock('react-native', () => {
//   return Object.setPrototypeOf(
//     {
//       Dimensions: {
//         get: () => ({width: 414, height: 896}),
//       },
//       Platform: {
//         ...ReactNative.Platform,
//         OS: 'ios',
//         select: jest.fn((selector) => selector),
//       },
//       NativeModules: {
//         ...ReactNative.NativeModules,
//         SettingsManager: {
//           settings: {
//             AppleLocale: 'en_US',
//           },
//         },
//         I18nManager: {
//           localeIdentifier: 'en_US',
//         },
//       },
//     },
//     ReactNative,
//   );
// });
