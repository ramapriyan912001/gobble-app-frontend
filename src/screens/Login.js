import React, {useState} from 'react'
import {Text, View, TextInput, Image, TouchableOpacity, SafeAreaView, Alert, KeyboardAvoidingView, TouchableWithoutFeedback, Keyboard, Platform } from 'react-native'
import {StatusBar} from 'expo-status-bar'
import {imageStyles, inputStyles, buttonStyles, containerStyles} from '../styles/LoginStyles'
import {styles} from '../styles/RegisterStyles'
import themes from '../styles/Themes';
import {API} from '../api'
import firebaseSvc from '../firebase/FirebaseSvc';
import BottomTabs from '../components/BottomTabs'
import * as Haptics from 'expo-haptics';
import { useColorScheme } from 'react-native-appearance';

/**
 * Page for user to log in!
 * 
 * @param {*} props Props from previous screen
 * @returns Login page render function
 */
export default function Login(props) {
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const colorScheme = useColorScheme();
    const isLight = colorScheme === 'light';
    
    // Login Success/Fail handlers
    const loginSuccess = (userCredential) => {
        console.log('login successful');
        const user = userCredential.user;
        // user.lastSeen === null
        // ? props.navigation.navigate('Welcome')
        // :
        props.navigation.replace('BottomTabs');
    };
    const loginFailed = (err) => {
        // const errorCode = err.code;
        const errorMessage = err.message;
        Alert.alert(errorMessage);
    };

    /**
     * Login to user account. Uses the email password and user details from function state.
     * Calls firebaseSvc.login() to log in the user
     */
    const onPressLogin = () => {
        const user = {
        email: email,
        password: password,
        };
        
        firebaseSvc
        .login(user, loginSuccess, loginFailed);
    };

    const imageMargin = Platform.OS == 'ios' ? '7%':'2%';

    return(
        <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
            <SafeAreaView style={[themes.containerTheme(isLight), {flex:1, paddingTop:'20%'}]}>    
                <KeyboardAvoidingView 
                    enabled={true}
                    behavior={"position"}
                    >
                    <Image style={{...imageStyles.gobbleImage, marginTop:imageMargin,height:'40%',marginLeft:'20%'}} source = {require('../images/gobble.png')} testID={'GobbleImage'}/>
                    <StatusBar style="auto"/>
                    <View style={[{...styles.inputView, marginLeft:'20%'}, themes.viewTheme(isLight)]}> 
                        <TextInput
                            autoCapitalize="none"
                            style={[inputStyles.TextInput, themes.textTheme(isLight)]}
                            placeholder="Email"
                            placeholderTextColor={themes.oppositeTheme(isLight)}
                            returnKeyType = 'done'
                            onSubmitEditing={(event) => onPressLogin()}
                            onChangeText={(email) => setEmail(email)}
                        />
                    </View>
                        
                    <View style={[{...styles.inputView, marginLeft:'20%'}, themes.viewTheme(isLight)]}>
                        <TextInput
                            autoCapitalize="none"
                            style={[inputStyles.TextInput, themes.textTheme(isLight)]}
                            placeholder="Password"
                            placeholderTextColor={themes.oppositeTheme(isLight)}
                            returnKeyType = 'done'
                            secureTextEntry={true}
                            onSubmitEditing={(event) => onPressLogin()}
                            onChangeText={(password) => setPassword(password)}
                        />
                    </View>
                    <TouchableOpacity
                        style={{...buttonStyles.forgotButton, marginLeft:'30%'}}
                        onPress={()=> {
                        Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Small);
                        props.navigation.navigate('ForgotPassword');}}
                        testID={'ToForgotPasswordButton'}
                    >
                        <Text style={[buttonStyles.forgotButtonText, themes.textTheme(isLight)]}>Forgot Password?</Text>
                    </TouchableOpacity>
                    <TouchableOpacity
                        style={[styles.longButton, themes.buttonTheme(isLight)]}
                        onPress={() => {
                        Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Small);
                        onPressLogin();
                        }}
                        testID={'LoginButton'}
                    >
                        <Text style={[buttonStyles.loginButtonText, themes.oppositeTextTheme(isLight)]}>Log In</Text>
                    </TouchableOpacity>
                    <TouchableOpacity
                        style={[styles.longButton, themes.buttonTheme(isLight)]}
                        onPress={()=> {
                            Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Small);
                            props.navigation.navigate('Welcome');}}
                        testID={'ToWelcomeButton'}
                    >
                        <Text style={[buttonStyles.loginButtonText, themes.oppositeTextTheme(isLight)]}>Back</Text>
                    </TouchableOpacity>
                </KeyboardAvoidingView>
            </SafeAreaView>
        </TouchableWithoutFeedback>
    )
}
/**
 * 
 */