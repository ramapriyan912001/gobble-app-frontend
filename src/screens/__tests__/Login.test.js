import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Login from '../Login';
import renderer from 'react-test-renderer';
import {jest, expect} from '@jest/globals';

describe('<Login />', () => {
    // it('should render correctly', () => {
    //     const tree = renderer.create(<Login />).toJSON();
    //     expect(tree).toMatchSnapshot();
    //   });

    let tree;

    beforeEach(() => {
        tree = render(<Login/>)
    })

    test("snapshot", () => {
        expect(tree).toMatchSnapshot();
    })

    test("renders default elements", () => {
        //Act
        const localTree = render(<Login/>);
        const { getByPlaceholderText, getByText, getByTestId, findByTestId } = localTree;

        //Assert
        expect(findByTestId('GobbleImage')).not.toThrow();
        getByPlaceholderText('Email');
        getByPlaceholderText('Password');
        getByTestId('ToForgotPasswordButton');
        getByText('Forgot Password?');
        getByTestId('LoginButton');
        getByText('Log In');
        getByTestId('ToWelcomeButton');
        getByText('Back');
    });


    it('can navigate to ForgotPassword', () => {
        const navigateMock = jest.fn();
        const { getByTestId } = render(<Login navigation={{ navigate: navigateMock }}/>);
        fireEvent.press(getByTestId('ToForgotPasswordButton'));
        expect(navigateMock).toBeCalledWith('ForgotPassword');
    });

    it('can navigate to welcome', () => {
        const navigateMock = jest.fn();
        const { getByTestId } = render(<Login navigation={{ navigate: navigateMock }}/>);

        fireEvent.press(getByTestId('ToWelcomeButton'));
        expect(navigateMock).toBeCalledWith('Welcome');
    });
});