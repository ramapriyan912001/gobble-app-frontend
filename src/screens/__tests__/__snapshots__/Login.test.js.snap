// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> snapshot 1`] = `
<View
  onLayout={[Function]}
  scrollEnabled={false}
>
  <Image
    source={1}
    style={
      Object {
        "height": "30%",
        "marginBottom": "10%",
        "marginLeft": "2%",
        "marginTop": "-5%",
        "width": "60%",
      }
    }
    testID="GobbleImage"
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 30,
          "elevation": 5,
          "height": 45,
          "marginBottom": 20,
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "width": "60%",
        },
        Object {
          "backgroundColor": "#242C40",
          "shadowColor": "#ffe",
          "shadowRadius": 0.1,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Email"
      placeholderTextColor="#ffe"
      rejectResponderTermination={true}
      returnKeyType="done"
      style={
        Array [
          Object {
            "flex": 1,
            "height": "5%",
            "paddingLeft": "6%",
            "width": 200,
          },
          Object {
            "color": "#ffe",
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 30,
          "elevation": 5,
          "height": 45,
          "marginBottom": 20,
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "width": "60%",
        },
        Object {
          "backgroundColor": "#242C40",
          "shadowColor": "#ffe",
          "shadowRadius": 0.1,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Password"
      placeholderTextColor="#ffe"
      rejectResponderTermination={true}
      returnKeyType="done"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "flex": 1,
            "height": "5%",
            "paddingLeft": "6%",
            "width": 200,
          },
          Object {
            "color": "#ffe",
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 30,
        "marginBottom": "6%",
        "opacity": 1,
      }
    }
    testID="ToForgotPasswordButton"
  >
    <Text
      style={
        Array [
          Object {},
          Object {
            "color": "#ffe",
          },
        ]
      }
    >
      Forgot Password?
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#ffe",
        "borderRadius": 25,
        "elevation": 5,
        "height": 50,
        "justifyContent": "center",
        "marginTop": "5%",
        "opacity": 1,
        "shadowColor": "#ffe",
        "shadowOffset": Object {
          "height": 4,
          "width": 4,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 0.1,
        "width": "85%",
      }
    }
    testID="LoginButton"
  >
    <Text
      style={
        Array [
          Object {},
          Object {
            "color": "#242C40",
          },
        ]
      }
    >
      Log In
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#ffe",
        "borderRadius": 25,
        "elevation": 5,
        "height": 50,
        "justifyContent": "center",
        "marginTop": "5%",
        "opacity": 1,
        "shadowColor": "#ffe",
        "shadowOffset": Object {
          "height": 4,
          "width": 4,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 0.1,
        "width": "85%",
      }
    }
    testID="ToWelcomeButton"
  >
    <Text
      style={
        Array [
          Object {},
          Object {
            "color": "#242C40",
          },
        ]
      }
    >
      Back
    </Text>
  </View>
</View>
`;
